src/iceberg/type.cc

```diff
@@ -0,0 +1,312 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "iceberg/type.h"
+
+#include <format>
+#include <iterator>
+#include <stdexcept>
+
+#include "iceberg/util/formatter.h"
+
+namespace iceberg {
+
+StructType::StructType(std::vector<SchemaField> fields) : fields_(std::move(fields)) {
+  size_t index = 0;
+  for (const auto& field : fields_) {
+    auto [it, inserted] = field_id_to_index_.try_emplace(field.field_id(), index);
+    if (!inserted) {
+      throw std::runtime_error(
+          std::format("StructType: duplicate field ID {} (field indices {} and {})",
+                      field.field_id(), it->second, index));
+    }
+
+    ++index;
+  }
+}
+
+TypeId StructType::type_id() const { return TypeId::kStruct; }
+std::string StructType::ToString() const {
+  std::string repr = "struct<\n";
```

Comment:
Is there a specific reason to use a separate line for each sub-field? An alternative is to use the Hive-style where all fields are concatenated in a single line.

Reply:
That gets messy IMO once there are more than a couple fields

Reply:
Do we want to support restoring a Schema object by deserializing the formatted string?

Reply:
I'd rather add JSON serializers than try to parse the repr